combinetf2/fitter.py

```diff
@@ -179,6 +183,79 @@ def __init__(self, indata, options):
             and ((not self.binByBinStat) or self.binByBinStatType == "normal")
         )
 
+    def set_blinding(self, blind=True, unblind_parameter_expressions=[]):
+        self.blind = blind
+
+        def compile_patterns(patterns):
+            compiled = []
+            for p in patterns:
+                if p.startswith("r:"):
+                    # Treat as regex, remove prefix
+                    compiled.append(re.compile(p[2:]))
+                else:
+                    # Treat as exact string match
+                    compiled.append(re.compile(rf"^{re.escape(p)}$"))
+            return compiled
+
+        # Find parameters that match any regex
+        compiled_regexes = compile_patterns(unblind_parameter_expressions)
+        unblind_parameters = [
+            s
+            for s in [*self.indata.procs, *self.indata.noigroups]
+            if any(regex.search(s.decode()) for regex in compiled_regexes)
+        ]
+
+        self.unblind_parameters = unblind_parameters
+
+    def apply_blinding_offsets(
+        self,
+        x,
+    ):
+        # random blinding with seed taken based on string of parameter name
+
+        # check if dataset is an integer (i.e. if it is real data or not) and use this to choose the random seed
+        is_dataobs_int = np.sum(
+            np.equal(self.indata.data_obs, np.floor(self.indata.data_obs))
+        )
+
+        def deterministic_random_from_string(s, mean=0.0, std=5.0):
+            if isinstance(s, str):
+                s = s.encode("utf-8")
+
+            if is_dataobs_int:
+                s += b"_data"
+
+            # Hash the string
+            hash = hashlib.sha256(s).hexdigest()
+
+            seed_seq = np.random.SeedSequence(int(hash, 16))
+            rng = np.random.default_rng(seed_seq)
+
+            value = rng.normal(loc=mean, scale=std)
+            return value
+
+        x_mult = np.ones(self.indata.nsyst + self.npoi, dtype=np.float64)
+        x_add = np.zeros_like(x_mult)
+
+        for i in range(self.npoi):
+            param = self.indata.procs[i]
+            if param in self.unblind_parameters:
+                continue
+            value = deterministic_random_from_string(param)
+            x_mult[i] = np.exp(value)
+
+        for i in self.indata.noigroupidxs:
+            param = self.indata.noigroups[i]
+            if param in self.unblind_parameters:
+                continue
+            value = deterministic_random_from_string(param)
+            x_add[i + self.npoi] = value
+
+        x = x * tf.constant(x_mult, dtype=self.indata.dtype)
```

Comment:
Should this have different treatment for the case where the poi's are squared?  (or perhaps be applied after the corresponding transformation?)

Reply:
Blinding is applied now after transformation in dedicated `get_poi()` function